File: services/risk_classifier/src/classifier.rs
use std::{cmp::Ordering, sync::Arc};

use serde::Serialize;

use crate::config::{RiskConfig, SpendThreshold};

#[derive(Debug, Clone)]
pub struct RiskClassifier {
    config: Arc<RiskConfig>,
}

impl RiskClassifier {
    #[must_use]
    pub fn new(config: RiskConfig) -> Self {
        Self {
            config: Arc::new(config),
        }
    }

    #[must_use]
    pub fn classify(&self, input: &RiskInput) -> RiskVerdict {
        let mut level = RiskLevel::Low;
        let mut score = 0.0;
        let mut reasons = Vec::new();

        if let Some(spend) = &input.spend {
            let currency = spend.currency.to_uppercase();
            if let Some(threshold) = self.config.spend_thresholds.get(&currency) {
                let normalized = threshold_normalized(threshold);
                let amount = spend.amount_minor_units;
                match amount.cmp(&normalized.high_minor_units) {
                    Ordering::Greater | Ordering::Equal => {
                        level = RiskLevel::High;
                        score += 0.4;
                        reasons.push(format!(
                            "amount {} {} exceeds high threshold {}",
                            amount, currency, normalized.high_minor_units
                        ));
                    }
                    Ordering::Less => {
                        if amount >= normalized.caution_minor_units {
                            level = level.max(RiskLevel::Medium);
                            score += 0.25;
                            reasons.push(format!(
                                "amount {} {} exceeds caution threshold {}",
                                amount, currency, normalized.caution_minor_units
                            ));
                        }
                    }
                }
            } else {
                reasons.push(format!(
                    "no spend threshold configured for {}; using baseline",
                    currency
                ));
                score += 0.05;
            }

            if spend
                .merchant
                .as_deref()
                .map(|merchant| merchant.to_ascii_lowercase().contains("crypto"))
                .unwrap_or(false)
            {
                level = level.max(RiskLevel::Medium);
                score += 0.2;
                reasons.push("merchant contains crypto keyword".into());
            }
        }

        let mut tag_score = 0.0;
        for tag in &input.tags {
            if let Some(weight) = self.config.tag_weights.get(tag) {
                tag_score += weight;
                reasons.push(format!("tag {tag} contributes {weight:.2} risk weight"));
            }
        }

        if tag_score >= self.config.tag_high_threshold {
            level = RiskLevel::High;
        } else if tag_score >= self.config.tag_medium_threshold {
            level = level.max(RiskLevel::Medium);
        }

        score += tag_score;

        let mut confidence = (self.config.baseline_confidence + score).clamp(0.05, 0.99);
        confidence = match level {
            RiskLevel::Low => confidence.min(0.6),
            RiskLevel::Medium => confidence.clamp(0.55, 0.8),
            RiskLevel::High => confidence.max(0.8),
        };

        RiskVerdict {
            level,
            confidence,
            reasons,
        }
    }
}

fn threshold_normalized(threshold: &SpendThreshold) -> SpendThreshold {
    threshold.normalized()
}

#[derive(Debug, Clone, Default)]
pub struct RiskInput {
    pub tags: Vec<String>,
    pub spend: Option<SpendContext>,
}

#[derive(Debug, Clone)]
pub struct SpendContext {
    pub amount_minor_units: u64,
    pub currency: String,
    pub merchant: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
#[repr(u8)]
pub enum RiskLevel {
    Low = 0,
    Medium = 1,
    High = 2,
}

impl RiskLevel {
    fn max(self, other: Self) -> Self {
        if (self as u8) >= (other as u8) {
            self
        } else {
            other
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct RiskVerdict {
    pub level: RiskLevel,
    pub confidence: f32,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub reasons: Vec<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::RiskConfig;

    #[test]
    fn upgrades_to_high_when_amount_crosses_threshold() {
        let mut config = RiskConfig::default();
        config.spend_thresholds.insert(
            "USD".into(),
            SpendThreshold {
                caution_minor_units: 50_000,
                high_minor_units: 100_000,
            },
        );
        let classifier = RiskClassifier::new(config);

        let input = RiskInput {
            tags: vec![],
            spend: Some(SpendContext {
                amount_minor_units: 120_000,
                currency: "usd".into(),
                merchant: Some("Coffee Club".into()),
            }),
        };

        let verdict = classifier.classify(&input);
        assert_eq!(verdict.level, RiskLevel::High);
        assert!(verdict.confidence >= 0.8);
    }

    #[test]
    fn considers_tag_weights_for_medium_risk() {
        let mut config = RiskConfig::default();
        config.tag_weights.insert("risk:manual_review".into(), 0.35);
        let classifier = RiskClassifier::new(config);

        let verdict = classifier.classify(&RiskInput {
            tags: vec!["risk:manual_review".into()],
            spend: None,
        });

        assert_eq!(verdict.level, RiskLevel::Medium);
        assert!(verdict.confidence >= 0.55);
        assert!(
            verdict
                .reasons
                .iter()
                .any(|reason| reason.contains("risk:manual_review"))
        );
    }
}

File: packages/gateway/src/modules/risk/classifier.ts
/**
 * Risk classifier — port of services/risk_classifier/src/classifier.rs
 * and services/risk_classifier/src/config.rs
 *
 * Pure risk scoring engine using configurable spend thresholds and tag weights.
 */

import type {
  RiskConfig,
  RiskInput,
  RiskVerdict,
  RiskLevel,
  SpendThreshold,
} from "@tyrum/schemas";

// ---------------------------------------------------------------------------
// Config helpers
// ---------------------------------------------------------------------------

export function defaultRiskConfig(): RiskConfig {
  return {
    baseline_confidence: 0.35,
    tag_medium_threshold: 0.3,
    tag_high_threshold: 0.6,
    tag_weights: {},
    spend_thresholds: {},
  };
}

function normalizeThreshold(threshold: SpendThreshold): SpendThreshold {
  const caution = Math.min(
    threshold.caution_minor_units,
    threshold.high_minor_units,
  );
  const high = Math.max(
    threshold.caution_minor_units,
    threshold.high_minor_units,
  );
  return { caution_minor_units: caution, high_minor_units: high };
}

export function normalizeConfig(config: RiskConfig): RiskConfig {
  let tagMedium = config.tag_medium_threshold;
  let tagHigh = config.tag_high_threshold;

  if (tagHigh < tagMedium) {
    const tmp = tagHigh;
    tagHigh = tagMedium;
    tagMedium = tmp;
  }

  const normalizedSpend: Record<string, SpendThreshold> = {};
  for (const [currency, threshold] of Object.entries(config.spend_thresholds)) {
    normalizedSpend[currency] = normalizeThreshold(threshold);
  }

  return {
    baseline_confidence: config.baseline_confidence,
    tag_medium_threshold: tagMedium,
    tag_high_threshold: tagHigh,
    tag_weights: { ...config.tag_weights },
    spend_thresholds: normalizedSpend,
  };
}

// ---------------------------------------------------------------------------
// Risk level ordering
// ---------------------------------------------------------------------------

const RISK_ORDER: Record<RiskLevel, number> = {
  low: 0,
  medium: 1,
  high: 2,
};

function maxRiskLevel(a: RiskLevel, b: RiskLevel): RiskLevel {
  const aScore = RISK_ORDER[a] ?? 0;
  const bScore = RISK_ORDER[b] ?? 0;
  return aScore >= bScore ? a : b;
}

// ---------------------------------------------------------------------------
// Classifier
// ---------------------------------------------------------------------------

export class RiskClassifier {
  private config: RiskConfig;

  constructor(config: RiskConfig) {
    this.config = normalizeConfig(config);
  }

  classify(input: RiskInput): RiskVerdict {
    let level: RiskLevel = "low";
    let score = 0;
    const reasons: string[] = [];

    if (input.spend != null) {
      const spend = input.spend;
      const currency = spend.currency.toUpperCase();
      const threshold = this.config.spend_thresholds[currency];

      if (threshold != null) {
        const normalized = normalizeThreshold(threshold);
        const amount = spend.amount_minor_units;

        if (amount >= normalized.high_minor_units) {
          level = "high";
          score += 0.4;
          reasons.push(
            `amount ${amount} ${currency} exceeds high threshold ${normalized.high_minor_units}`,
          );
        } else if (amount >= normalized.caution_minor_units) {
          level = maxRiskLevel(level, "medium");
          score += 0.25;
          reasons.push(
            `amount ${amount} ${currency} exceeds caution threshold ${normalized.caution_minor_units}`,
          );
        }
      } else {
        reasons.push(
          `no spend threshold configured for ${currency}; using baseline`,
        );
        score += 0.05;
      }

      if (
        spend.merchant != null &&
        spend.merchant.toLowerCase().includes("crypto")
      ) {
        level = maxRiskLevel(level, "medium");
        score += 0.2;
        reasons.push("merchant contains crypto keyword");
      }
    }

    let tagScore = 0;
    for (const tag of input.tags) {
      const weight = this.config.tag_weights[tag];
      if (weight != null) {
        tagScore += weight;
        reasons.push(
          `tag ${tag} contributes ${weight.toFixed(2)} risk weight`,
        );
      }
    }

    if (tagScore >= this.config.tag_high_threshold) {
      level = "high";
    } else if (tagScore >= this.config.tag_medium_threshold) {
      level = maxRiskLevel(level, "medium");
    }

    score += tagScore;

    let confidence = Math.min(
      Math.max(this.config.baseline_confidence + score, 0.05),
      0.99,
    );
    switch (level) {
      case "low":
        confidence = Math.min(confidence, 0.6);
        break;
      case "medium":
        confidence = Math.min(Math.max(confidence, 0.55), 0.8);
        break;
      case "high":
        confidence = Math.max(confidence, 0.8);
        break;
    }

    return { level, confidence, reasons };
  }
}
